hadoop-hdds/docs/content/design/container-reconciliation.md

```diff
@@ -0,0 +1,246 @@
+      - Null handling should still be in place for containers created before this feature, and to guard against bugs.
+
+#### On Container Scan
+
+1. Scanner reads chunks and checks checksums in RocksDB.
+2. Scanner updates Merkle tree to match the current contents of the disk. It does not update RocksDB.
+    - Update Merkle tree checksums in a rolling fashion per chunk and block as scanning progresses.
+    -  Do not let the scanner overwrite the RocksDB value, otherwise we cannot locally detect corruption anymore.
+        - This is not a deal breaker since SCM would see it on the next full container report, but it may slow down bitrot detection.
+3. If scanner detects local corruption (RocksDB hash does not match hash calculated from disk), marks container unhealthy and sends an incrmeental container report to SCM with its calculated container hash.
+    - The Merkle Tree is still updated to reflect the current contents of the disk in this case.
+
+#### On Container import
+
+- Schedule on-demand scan after the container import completes.
+    - There may be cases where we need to import a slightly corrupt container since it is our best option. Therefore, do not do a synchronous scan of container and fail the import if corruption is detected.
+    - Instead, schedule the imported container for on-demand scanning to be sure SCM knows of corruption if it does not already.
+    - **Optimization**: Clear last scanned timestamp on container import to make the background scanner get to it faster if datanode restarts before the on-demand scan completes.
+
+#### On `getContainerHashes`
+
+- Return the current Merkle Tree of the specified container.
+    - The tree will be calculated from existing chunk checksums if it is not present. However, this should not happen in the majority of cases since it will be calculated on container close.
+- If container not closed, return error.
+- If container not found, return error.
+
+- The merkle tree returned by this call will always be consistent.
+    - The hash of each node in the tree will always accurately represent a hash of its children.
+- The merkle tree returned by this call may not always match the current state of the container.
+    - The container can be repaired or corrupted any time between the scan that generated the tree and the call to read the container's tree.
+    - If the caller decides it needs to read a chunk based on the merkle tree obtained, it will have to re-check the checksum obtained as part of the chunk read to verify the data received matches what it was expecting.
+
+#### On `reconcileContainer`
+
+- Use existing merkle tree, unless its not there, then compute it.
+    - The tree will be calculated from existing chunk checksums if it is not present. However, this should not happen in the majority of cases since it will be calculated on container close.
+1. For each replica:
```

Comment:
What about finding missed deletes? Is that something we want to address too?

In the case of a block missing from one replica (or an additional block in one replica). MTs are ordered structures, so if you are missing a block or have an extra block in one MT "early" in the sorted order, it will effectively make all the hashes differ between the two trees and we will need to fall back to a block by block comparison. How do we handle that, or is my understanding incorrect?

Reply:
Yes there would be a linear walk of the blocks to reconcile which is not different from a width first walk of a tree.